include/insns/Zicsr.h

```diff
@@ -0,0 +1,184 @@
+  static uint64_t rdtimeImpl( RevFeature* F, T* R, RevMem* M, const RevInst& Inst ) {
+    return R->Core->GetCurrentSimCycle();
+  }
+
+  static uint64_t rdinstretImpl( RevFeature* F, RevRegFile* R, RevMem* M, const RevInst& Inst ) { return R->GetInstRet(); }
+
+  enum Half { Lo, Hi };
+
+  /// Performance Counter template
+  // Passed a function which gets the 64-bit value of a performance counter
+  template<Half HALF, uint64_t COUNTER( RevFeature* F, RevRegFile* R, RevMem* M, const RevInst& Inst )>
+  static bool perfCounter( RevFeature* F, RevRegFile* R, RevMem* M, const RevInst& Inst ) {
+    if( R->IsRV32 ) {
+      if constexpr( HALF == Lo ) {
+        R->SetX( Inst.rd, static_cast<uint32_t>( COUNTER( F, R, M, Inst ) & 0xffffffff ) );
+      } else {
+        R->SetX( Inst.rd, static_cast<uint32_t>( COUNTER( F, R, M, Inst ) >> 32 ) );
+      }
+    } else {
+      if constexpr( HALF == Lo ) {
+        R->SetX( Inst.rd, COUNTER( F, R, M, Inst ) );
+      } else {
+        return false;  // Hi half is not available on RV64
+      }
+    }
+    R->AdvancePC( Inst );
+    return true;
+  }
+
+  static constexpr auto& rdcycle    = perfCounter<Lo, rdcycleImpl>;
+  static constexpr auto& rdcycleh   = perfCounter<Hi, rdcycleImpl>;
+  static constexpr auto& rdtime     = perfCounter<Lo, rdtimeImpl>;
+  static constexpr auto& rdtimeh    = perfCounter<Hi, rdtimeImpl>;
+  static constexpr auto& rdinstret  = perfCounter<Lo, rdinstretImpl>;
+  static constexpr auto& rdinstreth = perfCounter<Hi, rdinstretImpl>;
+
+  // ----------------------------------------------------------------------
+  //
+  // RISC-V CSR Instructions
+  //
+  // ----------------------------------------------------------------------
+  struct RevZicsrInstDefaults : RevInstDefaults {
+    RevZicsrInstDefaults() {
+      SetOpcode( 0b1110011 );
+      SetRaiseFPE( true );
+      SetrdClass( RevRegClass::RegGPR );
+      Setrs2Class( RevRegClass::RegUNKNOWN );
+      Setimm( FImm );
+      SetFormat( RVTypeI );
+    }
+  };
+
+  // clang-format off
+  std::vector<RevInstEntry> ZicsrTable = {
+    { RevZicsrInstDefaults().SetMnemonic( "csrw %csr, %rs1"        ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) == 0; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "csrrw %csr, %rd, %rs1"  ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) != 0x1 && DECODE_IMM12( Inst ) != 0x2; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "fsflags %rs"            ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) == 0x1; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "fsrm %rs"               ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) == 0x2; } ) },
```

Comment:
`fsflags` and `fsrm` are pseudoinstructions which are `csrrw` instructions underneath.